File: src/tools/codegraph.ts
import * as path from "node:path";
import * as fs from "node:fs";
import { run, which } from "../util/exec.js";
import { log } from "../util/logger.js";
import { agentPaths, home } from "../util/paths.js";
import { npmGlobalInstall } from "../util/npm-install.js";
import { stringifyJson } from "../util/jsonc.js";
import { upsertBlock, removeBlock } from "../util/toml.js";
import type { ToolManifest, RunOpts } from "../core/tool-manifest.js";

async function ensureInstalled(opts: RunOpts): Promise<boolean> {
  if (process.env.TOKLESS_TEST === "1") {
    const dest = path.join(home(), ".local", "bin");
    fs.mkdirSync(dest, { recursive: true });
    const cgPath = path.join(dest, "codegraph");
    if (fs.existsSync(cgPath)) {
      try { fs.unlinkSync(cgPath); } catch { /* noop */ }
    }
    fs.writeFileSync(cgPath, "#!/bin/sh\necho ok", { mode: 0o755 });
    const sep = process.platform === "win32" ? ";" : ":";
    const curPath = process.env.PATH || "";
    if (!curPath.split(sep).includes(dest)) process.env.PATH = dest + sep + curPath;
    return true;
  }
  if (which("codegraph") && !opts.upgrade) return true;
  if (opts.dryRun) {
    log.sub("[dry-run] would install @colbymchenry/codegraph globally");
    return true;
  }
  const v = await npmGlobalInstall("@colbymchenry/codegraph", "latest");
  return v !== null;
}

// `codegraph install` configures every agent CodeGraph supports in one call.
// We invoke it once per tokless run and reuse the result for each agent.
let realInstallDone: Promise<boolean> | null = null;
function callRealInstall(opts: RunOpts): Promise<boolean> {
  if (opts.dryRun) {
    log.sub("[dry-run] would run: codegraph install --yes");
    return Promise.resolve(true);
  }
  if (!realInstallDone) {
    realInstallDone = (async () => {
      const helpRes = await run("codegraph", ["install", "--help"], { capture: true });
      const hasYes = helpRes.stdout.includes("--yes") || helpRes.stderr.includes("--yes");
      const args = ["install", ...(hasYes ? ["--yes"] : [])];
      const r = await run("codegraph", args, { capture: true });
      return r.code === 0;
    })();
  }
  return realInstallDone;
}

// Test-mode shim writes the same per-agent files real `codegraph install` would.
function testShim(agent: "claude" | "opencode" | "codex"): boolean {
  if (agent === "claude") {
    const cp = agentPaths.claudeCode();
    fs.mkdirSync(cp.dir, { recursive: true });
    const current = fs.existsSync(cp.globalJson) ? JSON.parse(fs.readFileSync(cp.globalJson, "utf8")) : {};
    current.mcpServers = current.mcpServers || {};
    current.mcpServers.codegraph = { type: "stdio", command: "codegraph", args: ["serve", "--mcp"] };
    fs.writeFileSync(cp.globalJson, stringifyJson(current), "utf8");
  } else if (agent === "opencode") {
    const op = agentPaths.opencode();
    fs.mkdirSync(op.dir, { recursive: true });
    const current = fs.existsSync(op.config) ? JSON.parse(fs.readFileSync(op.config, "utf8")) : {};
    current.mcp = current.mcp || {};
    current.mcp.codegraph = { type: "local", command: ["codegraph", "serve", "--mcp"], enabled: true };
    fs.writeFileSync(op.config, stringifyJson(current), "utf8");
  } else {
    const cx = agentPaths.codex();
    fs.mkdirSync(cx.dir, { recursive: true });
    const current = fs.existsSync(cx.config) ? fs.readFileSync(cx.config, "utf8") : "";
    const updated = upsertBlock(current, {
      header: "mcp_servers.codegraph",
      fields: { command: "codegraph", args: ["serve", "--mcp"] },
    });
    fs.writeFileSync(cx.config, updated, "utf8");
  }
  return true;
}

function wire(agent: "claude" | "opencode" | "codex") {
  return async (opts: RunOpts): Promise<boolean> => {
    if (process.env.TOKLESS_TEST === "1") return testShim(agent);
    return callRealInstall(opts);
  };
}

async function unwireClaude(_opts: RunOpts): Promise<boolean> {
  const cp = agentPaths.claudeCode();
  if (!fs.existsSync(cp.globalJson)) return true;
  try {
    const cfg = JSON.parse(fs.readFileSync(cp.globalJson, "utf8"));
    if (cfg?.mcpServers?.codegraph) {
      delete cfg.mcpServers.codegraph;
      fs.writeFileSync(cp.globalJson, stringifyJson(cfg), "utf8");
    }
  } catch { /* file corrupt, leave it */ }
  return true;
}
async function unwireOpenCode(_opts: RunOpts): Promise<boolean> {
  const op = agentPaths.opencode();
  if (!fs.existsSync(op.config)) return true;
  try {
    const cfg = JSON.parse(fs.readFileSync(op.config, "utf8"));
    if (cfg?.mcp?.codegraph) {
      delete cfg.mcp.codegraph;
      fs.writeFileSync(op.config, stringifyJson(cfg), "utf8");
    }
  } catch { /* noop */ }
  return true;
}
async function unwireCodex(_opts: RunOpts): Promise<boolean> {
  const cx = agentPaths.codex();
  if (!fs.existsSync(cx.config)) return true;
  const raw = fs.readFileSync(cx.config, "utf8");
  const next = removeBlock(raw, "mcp_servers.codegraph");
  if (next !== raw) fs.writeFileSync(cx.config, next, "utf8");
  return true;
}

function verifyClaude(): boolean {
  const cp = agentPaths.claudeCode();
  if (!fs.existsSync(cp.globalJson)) return false;
  try {
    const cfg = JSON.parse(fs.readFileSync(cp.globalJson, "utf8"));
    return !!cfg?.mcpServers?.codegraph;
  } catch { return false; }
}
function verifyOpenCode(): boolean {
  const op = agentPaths.opencode();
  if (!fs.existsSync(op.config)) return false;
  try {
    const cfg = JSON.parse(fs.readFileSync(op.config, "utf8"));
    return !!cfg?.mcp?.codegraph;
  } catch { return false; }
}
function verifyCodex(): boolean {
  const cx = agentPaths.codex();
  if (!fs.existsSync(cx.config)) return false;
  const raw = fs.readFileSync(cx.config, "utf8");
  return raw.includes("[mcp_servers.codegraph]");
}

const codegraph: ToolManifest = {
  id: "codegraph",
  label: "CodeGraph",
  description: "MCP server that lets agents query a code knowledge graph instead of reading raw files.",
  homepage: "https://github.com/colbymchenry/codegraph",
  installHint: "npm i -g @colbymchenry/codegraph",
  channel: "npm",

  install: ensureInstalled,
  wireFor: {
    claude: wire("claude"),
    opencode: wire("opencode"),
    codex: wire("codex"),
  },
  unwireFor: {
    claude: unwireClaude,
    opencode: unwireOpenCode,
    codex: unwireCodex,
  },
  verifyFor: {
    claude: verifyClaude,
    opencode: verifyOpenCode,
    codex: verifyCodex,
  },
};

export default codegraph;

File: internal/tools/codegraph.go
package tools

import (
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/HoangP8/tokless/internal/agents"
	"github.com/HoangP8/tokless/internal/core"
	"github.com/HoangP8/tokless/internal/util"
)

func codegraphEnsureInstalled(opts core.RunOpts) (bool, error) {
	if isTest() {
		return true, nil
	}
	opts.Reportf("checking", 0.1)
	if util.ResolveCodegraphBin() != "" && !opts.Upgrade {
		opts.Reportf("already installed", 1)
		return true, nil
	}
	if opts.DryRun {
		util.L.Sub("[dry-run] would install @colbymchenry/codegraph globally")
		return true, nil
	}
	opts.Reportf("npm install -g", 0.4)
	_, ok, _ := util.NpmGlobalInstall("@colbymchenry/codegraph", "latest")
	ok = ok && util.ResolveCodegraphBin() != ""
	if ok {
		opts.Reportf("ready", 1)
	}
	return ok, nil
}

var (
	realInstallOnce sync.Once
	realInstallRes  bool
)

// codegraphRealInstall runs `codegraph install --target <agent>` per call.
func codegraphRealInstall(opts core.RunOpts, agent string) bool {
	if opts.DryRun {
		util.L.Sub("[dry-run] would run: codegraph install --yes")
		return true
	}
	bin := util.ResolveCodegraphBin()
	if bin == "" {
		return false
	}
	help := util.Run(bin, []string{"install", "--help"}, util.RunOptions{Capture: true})
	hasYes := strings.Contains(help.Stdout, "--yes") || strings.Contains(help.Stderr, "--yes")
	hasTarget := strings.Contains(help.Stdout, "--target") || strings.Contains(help.Stderr, "--target")
	args := []string{"install"}
	if hasYes {
		args = append(args, "--yes")
	}
	if hasTarget {
		target := agent
		if target == "antigravity" {
			target = "gemini"
		}
		if target == "" {
			target = "all"
		}
		args = append(args, "--target", target)
	}
	return util.Run(bin, args, util.RunOptions{Capture: true}).Code == 0
}

// codegraphConfigureMcp writes the MCP entry tokless-side.
func codegraphConfigureMcp(agent string) bool {
	switch agent {
	case "claude":
		agents.ConfigureClaudeMcp("codegraph")
	case "opencode":
		agents.ConfigureOpenCodeMcp("codegraph")
	case "codex":
		agents.ConfigureCodexMcp("codegraph")
	case "antigravity":
		agents.ConfigureAntigravityMcp("codegraph")
	}
	return true
}

func codegraphVerify(agent string) bool {
	switch agent {
	case "claude":
		cp := util.ClaudeCodePaths()
		raw, ok := util.ReadFileSafe(cp.GlobalJSON)
		if !ok {
			return false
		}
		cfg := util.TryParseJsonc(raw)
		if cfg == nil {
			return false
		}
		if s, ok := cfg.Get("mcpServers"); ok {
			if sm, ok := s.(*util.OrderedMap); ok {
				_, has := sm.Get("codegraph")
				return has
			}
		}
		return false
	case "opencode":
		op := util.OpenCodePathsResolved()
		raw, ok := util.ReadFileSafe(op.Config)
		if !ok {
			return false
		}
		cfg := util.TryParseJsonc(raw)
		if cfg == nil {
			return false
		}
		if m, ok := cfg.Get("mcp"); ok {
			if mm, ok := m.(*util.OrderedMap); ok {
				_, has := mm.Get("codegraph")
				return has
			}
		}
		return false
	case "codex":
		cx := util.CodexPathsResolved()
		raw, _ := util.ReadFileSafe(cx.Config)
		return strings.Contains(raw, "[mcp_servers.codegraph]")
	case "antigravity":
		agents.CleanupDeadIdeHooks()
		return agents.AntigravityMcpHas("codegraph") && agents.HasAntigravityCodegraphIndexHook()
	}
	return false
}

func codegraphIndexed(dir string, _ core.RunOpts) bool {
	return util.Exists(filepath.Join(dir, ".codegraph"))
}

func codegraphIndexProject(dir string, opts core.RunOpts) (bool, error) {
	if isTest() {
		_ = os.MkdirAll(filepath.Join(dir, ".codegraph"), 0o755)
		return true, nil
	}
	bin := util.ResolveCodegraphBin()
	if bin == "" {
		return false, nil
	}
	if opts.DryRun {
		util.L.Sub("[dry-run] would run codegraph in " + dir)
		return true, nil
	}
	go codegraphSyncBackground(bin, dir)
	return true, nil
}

func codegraphSyncBackground(bin, dir string) {
	if util.Exists(filepath.Join(dir, ".codegraph")) {
		_ = util.Run(bin, []string{"sync"}, util.RunOptions{Cwd: dir, Quiet: true})
		return
	}
	if util.Run(bin, []string{"init", "-i"}, util.RunOptions{Cwd: dir, Quiet: true}).Code != 0 {
		_ = util.Run(bin, []string{"init"}, util.RunOptions{Cwd: dir, Quiet: true})
	}
}

func codegraphWire(agent string) core.AgentFn {
	return func(opts core.RunOpts) (bool, error) {
		if isTest() {
			codegraphConfigureMcp(agent)
			WriteOwner(agent, "codegraph")
			if agent == "antigravity" {
				agents.InstallAntigravityCodegraphIndexHook()
				agents.RemoveAntigravityCodegraphToolDefs()
				agents.CleanupDeadIdeHooks()
			}
			return codegraphVerify(agent), nil
		}
		if opts.DryRun {
			return codegraphRealInstall(opts, agent), nil
		}
		if ran := codegraphRealInstall(opts, agent); !ran {
			util.L.Debug("codegraph's own installer failed; writing MCP entry directly")
		}
		codegraphConfigureMcp(agent)
		writeCodegraphBlock(agent)
		unwireAutoIndex(agent)
		if agent == "antigravity" {
			agents.InstallAntigravityCodegraphIndexHook()
			agents.RemoveAntigravityCodegraphToolDefs()
			agents.CleanupDeadIdeHooks()
		}
		return codegraphVerify(agent), nil
	}
}

// writeCodegraphBlock writes the unified TOKLESS block with codegraph as one
// of its owners.
func writeCodegraphBlock(agent string) bool {
	return WriteOwner(agent, "codegraph")
}

func unwireAutoIndex(agent string) {
	switch agent {
	case "claude":
		unwireClaudeAutoIndex()
	case "codex":
		unwireCodexAutoIndex()
	case "opencode":
		unwireOpencodeAutoIndex()
	case "antigravity":
		unwireGeminiAutoIndex()
	}
}

var codegraph = &core.ToolManifest{
	ID:           "codegraph",
	Label:        "CodeGraph",
	Description:  "MCP server that lets agents query a code knowledge graph instead of reading raw files.",
	Homepage:     "https://github.com/colbymchenry/codegraph",
	InstallHint:  "npm i -g @colbymchenry/codegraph",
	Channel:      core.ChannelNpm,
	Install:      codegraphEnsureInstalled,
	IndexProject: codegraphIndexProject,
	Indexed:      codegraphIndexed,
	IndexReady:   func() bool { return isTest() || util.ResolveCodegraphBin() != "" },
	WireFor: map[string]core.AgentFn{
		"claude":      codegraphWire("claude"),
		"opencode":    codegraphWire("opencode"),
		"codex":       codegraphWire("codex"),
		"antigravity": codegraphWire("antigravity"),
	},
	UnwireFor: map[string]core.AgentFn{
		"claude": func(core.RunOpts) (bool, error) {
			agents.RemoveClaudeMcp("codegraph")
			unwireAutoIndex("claude")
			RemoveOwner("claude", "codegraph")
			return true, nil
		},
		"opencode": func(core.RunOpts) (bool, error) {
			agents.RemoveOpenCodeMcp("codegraph")
			unwireAutoIndex("opencode")
			RemoveOwner("opencode", "codegraph")
			return true, nil
		},
		"codex": func(core.RunOpts) (bool, error) {
			cx := util.CodexPathsResolved()
			raw, ok := util.ReadFileSafe(cx.Config)
			if ok {
				next := util.RemoveBlock(raw, "mcp_servers.codegraph")
				if next != raw {
					_ = util.WriteFile(cx.Config, next)
				}
			}
			unwireAutoIndex("codex")
			RemoveOwner("codex", "codegraph")
			return true, nil
		},
		"antigravity": func(core.RunOpts) (bool, error) {
			agents.RemoveAntigravityMcp("codegraph")
			agents.RemoveAntigravityCodegraphIndexHook()
			unwireAutoIndex("antigravity")
			agents.CleanupDeadIdeHooks()
			agents.RemoveAntigravityCodegraphToolDefs()
			RemoveOwner("antigravity", "codegraph")
			return true, nil
		},
	},
	VerifyFor: map[string]core.VerifyFn{
		"claude":      func() *bool { return core.BoolPtr(codegraphVerify("claude")) },
		"opencode":    func() *bool { return core.BoolPtr(codegraphVerify("opencode")) },
		"codex":       func() *bool { return core.BoolPtr(codegraphVerify("codex")) },
		"antigravity": func() *bool { return core.BoolPtr(codegraphVerify("antigravity")) },
	},
}
